File: backend/v1/app/routers/categories.py
"""Category endpoints.

`GET /categories` is the questionnaire's data source and is open to every
signed-in user; everything that edits the tree is facility-admin only.

The tree is returned whole rather than paginated: it is five groups and about
thirty subcategories, and the report form needs all of it before it can render
step one.
"""

import uuid

from fastapi import APIRouter, status

from app.models.category import Category
from app.schemas.category import (
    CategoryCreate,
    CategoryNode,
    CategoryRead,
    CategoryTree,
    CategoryUpdate,
)
from app.schemas.common import DeleteResult
from app.security.dependencies import ADMIN_ONLY, SIGNED_IN, DbSession, IncludeInactive
from app.services import categories as service

router = APIRouter(prefix="/categories", tags=["categories"])


@router.get(
    "",
    response_model=CategoryTree,
    dependencies=[SIGNED_IN],
    summary="The category tree",
)
def get_category_tree(session: DbSession, include_inactive: IncludeInactive) -> CategoryTree:
    """Return every group with its subcategories nested inside."""
    groups = service.load_tree(session, include_inactive=include_inactive)
    return CategoryTree(groups=[_to_node(group) for group in groups])


@router.post(
    "",
    response_model=CategoryRead,
    status_code=status.HTTP_201_CREATED,
    dependencies=[ADMIN_ONLY],
    summary="Create a group or subcategory",
)
def create_category(payload: CategoryCreate, session: DbSession) -> CategoryRead:
    """Create a group, or a subcategory when `parent_id` names one.

    A `parent_id` that points at a subcategory is a 422: the tree is two levels
    deep and stays that way.
    """
    category = service.create_category(session, payload)
    session.commit()
    return CategoryRead.model_validate(category)


@router.get(
    "/{category_id}",
    response_model=CategoryRead,
    dependencies=[SIGNED_IN],
    summary="Get one category",
)
def get_category(category_id: uuid.UUID, session: DbSession) -> CategoryRead:
    """Return one group or subcategory."""
    return CategoryRead.model_validate(service.get_category(session, category_id))


@router.patch(
    "/{category_id}",
    response_model=CategoryRead,
    dependencies=[ADMIN_ONLY],
    summary="Update a category",
)
def update_category(
    category_id: uuid.UUID,
    payload: CategoryUpdate,
    session: DbSession,
) -> CategoryRead:
    """Rename, reorder, re-icon or deactivate a category.

    Changing a group's `location_detail` rewrites its subcategories to match.
    Setting a group-only field on a subcategory is a 422.
    """
    category = service.update_category(session, category_id, payload)
    session.commit()
    return CategoryRead.model_validate(category)


@router.delete(
    "/{category_id}",
    response_model=DeleteResult,
    dependencies=[ADMIN_ONLY],
    summary="Delete or deactivate a category",
)
def delete_category(category_id: uuid.UUID, session: DbSession) -> DeleteResult:
    """Remove a category, or deactivate it when it cannot be removed.

    A category incidents were filed under, and a group that still has
    subcategories, are deactivated instead; `deleted` and `deactivated` in the
    response say which happened.
    """
    result = service.delete_category(session, category_id)
    session.commit()
    return result


def _to_node(group: Category) -> CategoryNode:
    """Convert one eagerly-loaded group into its tree node."""
    return CategoryNode(
        **CategoryRead.model_validate(group).model_dump(),
        children=[CategoryRead.model_validate(child) for child in group.children],
    )

File: frontend/src/api/categories.ts
import { apiClient } from './client';
import type { Category, CategoryTree, DeleteResult, LocationDetail } from './types';

/**
 * The two-level category tree.
 *
 * A group is a category with no parent; a subcategory is one whose parent is a
 * group. There is no third level, and the API refuses to create one.
 */

/** Body of `POST /categories`. Omit `parent_id` to create a group. */
export interface CategoryCreatePayload {
  name: string;
  parent_id?: string | null;
  hint?: string | null;
  icon?: string | null;
  location_detail?: LocationDetail;
  sort_order?: number;
}

/**
 * Body of `PATCH /categories/{id}`.
 *
 * `parent_id` is deliberately absent: moving a subcategory between groups
 * would silently reclassify every incident already filed under it.
 */
export interface CategoryUpdatePayload {
  name?: string;
  hint?: string | null;
  icon?: string | null;
  location_detail?: LocationDetail;
  sort_order?: number;
  is_active?: boolean;
}

/**
 * Every group with its subcategories nested inside, in display order.
 *
 * One request rather than two: the questionnaire's first step renders the
 * groups and its second step renders `children`, and a user moves between them
 * in under a second.
 */
export async function fetchCategoryTree(includeInactive = false): Promise<CategoryTree> {
  const { data } = await apiClient.get<CategoryTree>('/categories', {
    params: includeInactive ? { include_inactive: true } : undefined,
  });
  return data;
}

/** Create a group or a subcategory. Admin only. */
export async function createCategory(payload: CategoryCreatePayload): Promise<Category> {
  const { data } = await apiClient.post<Category>('/categories', payload);
  return data;
}

/** Apply a partial update to a category. Admin only. */
export async function updateCategory(
  id: string,
  payload: CategoryUpdatePayload,
): Promise<Category> {
  const { data } = await apiClient.patch<Category>(`/categories/${id}`, payload);
  return data;
}

/** Delete a category, or deactivate it when incidents already reference it. */
export async function deleteCategory(id: string): Promise<DeleteResult> {
  const { data } = await apiClient.delete<DeleteResult>(`/categories/${id}`);
  return data;
}
